# 20.2 Support Vector Machines with Python

Welcome to the Support Vector Machines with Python Lecture Notebook! Remember to refer to the video lecture for the full background information on the code here!

## Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

## Get the Data

We'll use the built in breast cancer dataset from Scikit Learn. We can get with the load function:

In [2]:
from sklearn.datasets import load_breast_cancer

In [3]:
# Will be using the same data for PCA lectures
cancer = load_breast_cancer()

In [4]:
cancer.keys()

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])

In [5]:
print(cancer['DESCR']) # To print actual description of this dataset.

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry 
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 3 is Mean Radius, f

## Inferences from Description :
* There are 569 instances.
* 30 numeric attributes.
* Prediction that we want to do is whether or not, this tumor class is malignant or benign.
* Then we have data points like smoothness of tumor, concavity, symmetry, area etc.

## Setting up a dataframe using keys from cancer to actually grab the data

In [6]:
cancer['feature_names']

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

In [7]:
df_feat = pd.DataFrame(cancer['data'],columns=cancer['feature_names'])

In [8]:
df_feat.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

In [9]:
df_feat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
mean radius                569 non-null float64
mean texture               569 non-null float64
mean perimeter             569 non-null float64
mean area                  569 non-null float64
mean smoothness            569 non-null float64
mean compactness           569 non-null float64
mean concavity             569 non-null float64
mean concave points        569 non-null float64
mean symmetry              569 non-null float64
mean fractal dimension     569 non-null float64
radius error               569 non-null float64
texture error              569 non-null float64
perimeter error            569 non-null float64
area error                 569 non-null float64
smoothness error           569 non-null float64
compactness error          569 non-null float64
concavity error            569 non-null float64
concave points error       569 non-null float64
symmetry error             569 

In [10]:
# Target  
# cancer['target']

# Target names
cancer['target_names']

array(['malignant', 'benign'], dtype='<U9')

# Exploratory Data Analysis

We'll skip the Data Viz part for this lecture since there are so many features that are hard to interpret if you don't have domain knowledge of cancer or tumor cells. In your project you will have more to visualize for the data. So for now we are skipping ahead and focusing on running the ML model to predict target.

## Train Test Split

In [11]:
from sklearn.model_selection import train_test_split

In [12]:
X = df_feat
y = cancer['target']

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    test_size=0.30,
                                                    random_state=101)

# Train the Support Vector Classifier

In [14]:
from sklearn.svm import SVC
# Support Vector Classifier model
model = SVC(gamma="auto")

In [15]:
model.fit(X_train, y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [16]:
predictions = model.predict(X_test)

In [17]:
from sklearn.metrics import confusion_matrix, classification_report

In [18]:
print("CONFUSION MATRIX \n", confusion_matrix(y_test, predictions))
print("\n")
print("CLASSIFICATION REPORT \n", classification_report(y_test, predictions))

CONFUSION MATRIX 
 [[  0  66]
 [  0 105]]


CLASSIFICATION REPORT 
               precision    recall  f1-score   support

           0       0.00      0.00      0.00        66
           1       0.61      1.00      0.76       105

   micro avg       0.61      0.61      0.61       171
   macro avg       0.31      0.50      0.38       171
weighted avg       0.38      0.61      0.47       171



C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


- As we can see above on running the code above, basically what happened is that the model with default values predicted that no tumors were in zero class and we will get a warning that Precision and F-score are ill-defined and are being set to 0.0 in labels with no predicted samples.

- As, we can see it predicted everything belongging to class 1. It is because our model needs to have its parameters adjusted.

- Normalizing the data would also help, when we pass it into the SVM.

- We can search for best parameters using GridSearch

- GridSearch allows us to find the right parameters such as what c or gamma values to use.

- Finding those parameters at times can be a hectic task but here we can be a bit lazy and try bunch of combinations and see what works best.

- This idea of creating a grid of parameters and trying out all the best combinations is called a grid search.

- SkLearn has this functionality built-in.

# Gridsearch

Finding the right parameters (like what C or gamma values to use) is a tricky task! But luckily, we can be a little lazy and just try a bunch of combinations and see what works best! This idea of creating a 'grid' of parameters and just trying out all the possible combinations is called a Gridsearch, this method is common enough that Scikit-learn has this functionality built in with GridSearchCV! The CV stands for cross-validation which is the

GridSearchCV takes a dictionary that describes the parameters that should be tried and a model to train. The grid of parameters is defined as a dictionary, where the keys are the parameters and the values are the settings to be tested. 

In [19]:
from sklearn.model_selection import GridSearchCV
# Takes in a dictionary which describes the parameter that should be tried in a model to train.
# Grid of parameters is defined as a dictionary where keys are the parameters and values are list of settings to be tested.
param_grid = {'C':[0.1,1,10,100,1000],'gamma':[1,0.1,0.01,0.001,0.0001]}
# Calling the parameter C, gamma and values to test as key-value pair.
# So now we have param_grid which we will be feeding to GridSearchCV

## Various Parameter for Support Vector Classifier SVC
SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)
 
- More about C and gamma values in ISLR.

- C controls the cost of misclassification on the training data. A large C value gives low bias and high variance.  Low Bias because we penalise the cost of misclassification a lot for larger C values. With a lower C value we won't be penalising the cost that much and it gives us higher bias and lower variance in that case. Again, bias variance trade off plays a crucial role here.

- Gamma is the free parameter of radio base function (rbf). Small gamma means gaussian of large variance and technically we could say that a large gamma value is going to lead to a high bias and low variance. Large gamma implies variance is large and further it implies that support vector does not have a widespread influence.

In [20]:
grid = GridSearchCV(SVC(),param_grid,verbose=15,cv=4)
# Another great thing of GridSearchCV() is that it is a meta_estimator, it takes in a parameter, just like we did for SVC and
# creates a new estimator that behaves exactly the same.

In [21]:
grid.fit(X_train,y_train)

Fitting 4 folds for each of 25 candidates, totalling 100 fits
[CV] C=0.1, gamma=1 ..................................................
[CV] ....................... C=0.1, gamma=1, score=0.63, total=   0.0s
[CV] C=0.1, gamma=1 ..................................................
[CV] ....................... C=0.1, gamma=1, score=0.63, total=   0.0s
[CV] C=0.1, gamma=1 ..................................................
[CV] ......... C=0.1, gamma=1, score=0.6363636363636364, total=   0.0s
[CV] C=0.1, gamma=1 ..................................................
[CV] ......... C=0.1, gamma=1, score=0.6363636363636364, total=   0.0s
[CV] C=0.1, gamma=0.1 ................................................
[CV] ..................... C=0.1, gamma=0.1, score=0.63, total=   0.0s
[CV] C=0.1, gamma=0.1 ................................................
[CV] ..................... C=0.1, gamma=0.1, score=0.63, total=   0.0s
[CV] C=0.1, gamma=0.1 ................................................
[CV] ....... C=

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:    0.1s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:    0.1s remaining:    0.0s
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    0.1s remaining:    0.0s


[CV] ...... C=0.1, gamma=0.01, score=0.6363636363636364, total=   0.0s
[CV] C=0.1, gamma=0.01 ...............................................
[CV] ...... C=0.1, gamma=0.01, score=0.6363636363636364, total=   0.0s
[CV] C=0.1, gamma=0.001 ..............................................
[CV] ................... C=0.1, gamma=0.001, score=0.63, total=   0.0s
[CV] C=0.1, gamma=0.001 ..............................................
[CV] ................... C=0.1, gamma=0.001, score=0.63, total=   0.0s
[CV] C=0.1, gamma=0.001 ..............................................
[CV] ..... C=0.1, gamma=0.001, score=0.6363636363636364, total=   0.0s
[CV] C=0.1, gamma=0.001 ..............................................
[CV] ..... C=0.1, gamma=0.001, score=0.6363636363636364, total=   0.0s
[CV] C=0.1, gamma=0.0001 .............................................
[CV] .................. C=0.1, gamma=0.0001, score=0.89, total=   0.0s
[CV] C=0.1, gamma=0.0001 .............................................
[CV] .

[Parallel(n_jobs=1)]: Done  11 out of  11 | elapsed:    0.1s remaining:    0.0s
[Parallel(n_jobs=1)]: Done  12 out of  12 | elapsed:    0.1s remaining:    0.0s
[Parallel(n_jobs=1)]: Done  13 out of  13 | elapsed:    0.1s remaining:    0.0s
[Parallel(n_jobs=1)]: Done  14 out of  14 | elapsed:    0.1s remaining:    0.0s


[CV] ....................... C=1, gamma=0.1, score=0.63, total=   0.0s
[CV] C=1, gamma=0.1 ..................................................
[CV] ......... C=1, gamma=0.1, score=0.6363636363636364, total=   0.0s
[CV] C=1, gamma=0.1 ..................................................
[CV] ......... C=1, gamma=0.1, score=0.6363636363636364, total=   0.0s
[CV] C=1, gamma=0.01 .................................................
[CV] ...................... C=1, gamma=0.01, score=0.63, total=   0.0s
[CV] C=1, gamma=0.01 .................................................
[CV] ...................... C=1, gamma=0.01, score=0.63, total=   0.0s
[CV] C=1, gamma=0.01 .................................................
[CV] ........ C=1, gamma=0.01, score=0.6363636363636364, total=   0.0s
[CV] C=1, gamma=0.01 .................................................
[CV] ........ C=1, gamma=0.01, score=0.6363636363636364, total=   0.0s
[CV] C=1, gamma=0.001 ................................................
[CV] .

[CV] ........ C=1000, gamma=1, score=0.6363636363636364, total=   0.0s
[CV] C=1000, gamma=0.1 ...............................................
[CV] .................... C=1000, gamma=0.1, score=0.63, total=   0.0s
[CV] C=1000, gamma=0.1 ...............................................
[CV] .................... C=1000, gamma=0.1, score=0.63, total=   0.0s
[CV] C=1000, gamma=0.1 ...............................................
[CV] ...... C=1000, gamma=0.1, score=0.6363636363636364, total=   0.0s
[CV] C=1000, gamma=0.1 ...............................................
[CV] ...... C=1000, gamma=0.1, score=0.6363636363636364, total=   0.0s
[CV] C=1000, gamma=0.01 ..............................................
[CV] ................... C=1000, gamma=0.01, score=0.63, total=   0.0s
[CV] C=1000, gamma=0.01 ..............................................
[CV] ................... C=1000, gamma=0.01, score=0.63, total=   0.0s
[CV] C=1000, gamma=0.01 ..............................................
[CV] .

[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    1.4s finished


GridSearchCV(cv=4, error_score='raise-deprecating',
       estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='rbf', max_iter=-1, probability=False, random_state=None,
  shrinking=True, tol=0.001, verbose=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'C': [0.1, 1, 10, 100, 1000], 'gamma': [1, 0.1, 0.01, 0.001, 0.0001]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=15)

In [22]:
grid.best_params_ # To find the best parameters

{'C': 1, 'gamma': 0.0001}

In [23]:
grid.best_estimator_

SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma=0.0001, kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [24]:
grid.best_score_ # To get the best score.

0.9522613065326633

In [25]:
grid_predictions = grid.predict(X_test)

In [26]:
print("CONFUSION MATRIX \n", confusion_matrix(y_test, grid_predictions))
print("\n")
print("CLASSIFICATION REPORT \n", classification_report(y_test, grid_predictions))

CONFUSION MATRIX 
 [[ 59   7]
 [  4 101]]


CLASSIFICATION REPORT 
               precision    recall  f1-score   support

           0       0.94      0.89      0.91        66
           1       0.94      0.96      0.95       105

   micro avg       0.94      0.94      0.94       171
   macro avg       0.94      0.93      0.93       171
weighted avg       0.94      0.94      0.94       171



- The model above is performing really well against what we had earlier, which was just horrible if I'm being polite. It just guessed everything was in class 1.

- This is the power of using GridSearchCV and a support vector machine is undoubtedly one of those models where you have to do GridSearch and there is no other viable option to have an intuition over guessing optimal Gamma and C values.

- Grid Searches can take a lot of time, especially on larger datasets and we are trying lots of parameters, this sort of param_grid for grid search will take a long time.

- IRL make sure data is clean, go ahead and do grid search on a very small set of grids, 1 or 2 and then make sure it works correctly and after that set your full param_grid with as many default values as you want in your list, we run our grid and then retire for some time, break for lunch, take coffee, do something else, while GridSearch finishes.

- Since the data we have above is not enormous that is why GridSearch ran pretty smoothly, also depends on how fast your computer is.

## In a Nutshell : 
1. Import necessary packages
2. Access built-in breast_cancer dataset in SciKit-Learn
3. Load and put it in a dataframe
4. Visualize
5. Train test split
6. Call the model SVC from svm family of sklearn
7. Fit the model to training data.
8. Did the predictions and saw confusion matrix
9. Was predicting all as class 1, did horribly and classified everything into one class.
10. GridSearch your way through C and gamma parameters passing in as key-value pair.
11. Enter estimator object and param_grid dict in GridSearch  with verbose.
12. Do grid.fit() on training data. Let it run.
13. Grab best_params_ and best_estimator_
14. Call predictions and compare using classification and confusion_matrix,

# Great job!